core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java

```diff
@@ -7330,4 +7315,57 @@ private boolean isInternalRole(String domain) {
 
         return domain.equals("Internal") || domain.equals("Application");
     }
+
+    private String removeDomainIfNotApplicationRole(String roleName) {
+
+        String formattedRoleName;
+        if (roleName.contains(UserCoreConstants.DOMAIN_SEPARATOR)
+                && roleName.toLowerCase().startsWith(APPLICATION_DOMAIN.toLowerCase())) {
+            formattedRoleName = roleName;
+        } else {
+            formattedRoleName = UserCoreUtil.removeDomainFromName(roleName);
+        }
+        return formattedRoleName;
+    }
+
+    private void createHybridRole(String roleName, String[] userList, org.wso2.carbon.user.api.Permission[] permissions)
+            throws UserStoreException {
+
+        // It is possible that the adding role could already exists at the table. But if concurrent requests were made,
+        // it is possible that the adding role does not exists at this moment, but it still could exists at the
+        // moment when DB query is called.
+        try {
+            hybridRoleManager.addHybridRole(roleName, userList);
+        } catch (UserStoreException e) {
+            // In case of a unique constraint violation.
+            if (ERROR_CODE_ROLE_ALREADY_EXISTS.getCode().equals(e.getErrorCode())
+                    && hybridRoleManager.isExistingRole(roleName)) {
```

Comment:
No need to check  hybridRoleManager.isExistingRole(roleName)

Reply:
Addressed in f3182fdc6dde2e06d9b218d4d730da3b85886952.